### PR TITLE
Add time to CloudEvent automatically if not present

### PR DESCRIPTION
**Pull request by yash-nisar**
Closes https://github.com/dapr/dapr/issues/5137

Signed-off-by: Yash Nisar <yashnisar@microsoft.com>

# Description
Add time to CloudEvent automatically if not present

## Issue reference
https://github.com/dapr/dapr/issues/5137

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [x] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [x] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### Test evidence: 

Tested with https://github.com/dapr/components-contrib/pull/2123

#### Without custom timestamp:

Request:
```
dapr publish --publish-app-id order-processor --pubsub orderpubsub --topic orders --data '{"specversion" : "1.0", "type" : "com.dapr.cloudevent.sent", "source" : "testcloudeventspubsub", "subject" : "Cloud Events Test", "id" : "someCloudEventId", "datacontenttype" : "application/cloudevents+json", "data" : {"orderId": "100"}}'
```

Response:
```
INFO[0034] HTTP API Called: POST /v1.0/publish/orderpubsub/orders  app_id=order-processor instance=Yashs-MacBook-Pro.local scope=dapr.runtime.http-info type=log ver=edge
== APP == Dapr pub/sub is subscribed to: [{"pubsubname": "orderpubsub", "topic": "orders", "route": "orders"}]
== APP == Subscriber received : {'attributes': {'specversion': '1.0', 'id': 'someCloudEventId', 'source': 'testcloudeventspubsub', 'type': 'com.dapr.cloudevent.sent', 'datacontenttype': 'application/cloudevents+json', 'subject': 'Cloud Events Test', 'time': '2022-09-23T11:58:41-05:00', 'pubsubname': 'orderpubsub', 'topic': 'orders', 'traceid': '00-12a04ebe04ef75caabfbe51c4b124e1f-ebbc8f4aaba62480-01', 'traceparent': '00-12a04ebe04ef75caabfbe51c4b124e1f-ebbc8f4aaba62480-01', 'tracestate': ''}, 'data': {'orderId': '100'}}
```

Request:
```
curl -X POST [http://localhost:58257/v1.0/publish/orderpubsub/orders](http://localhost:58257/v1.0/publish/orderpubsub/orders) -H "Content-Type: application/cloudevents+json" -d '{"specversion" : "1.0", "type" : "com.dapr.cloudevent.sent", "source" : "testcloudeventspubsub", "subject" : "Cloud Events Test", "id" : "someCloudEventId", "datacontenttype" : "application/cloudevents+json", "data" : {"orderId": "100"}}'
```

Response:
```
INFO[0265] HTTP API Called: POST /v1.0/publish/orderpubsub/orders  app_id=order-processor instance=Yashs-MacBook-Pro.local scope=dapr.runtime.http-info type=log ver=edge
== APP == Subscriber received : {'attributes': {'specversion': '1.0', 'id': 'someCloudEventId', 'source': 'testcloudeventspubsub', 'type': 'com.dapr.cloudevent.sent', 'datacontenttype': 'application/cloudevents+json', 'subject': 'Cloud Events Test', 'time': '2022-09-23T12:02:32-05:00', 'pubsubname': 'orderpubsub', 'topic': 'orders', 'traceid': '00-556b723c238f7f1c49815cedd29b2981-5050fe399a05e7bf-01', 'traceparent': '00-556b723c238f7f1c49815cedd29b2981-5050fe399a05e7bf-01', 'tracestate': ''}, 'data': {'orderId': '100'}}
== APP == 127.0.0.1 - - [23/Sep/2022 12:02:32] "POST /orders HTTP/1.1" 200 -
```

#### With custom timestamp:

Request:
```
dapr publish --publish-app-id order-processor --pubsub orderpubsub --topic orders --data '{"specversion" : "1.0", "type" : "com.dapr.cloudevent.sent", "source" : "testcloudeventspubsub", "subject" : "Cloud Events Test", "id" : "someCloudEventId", "time" : "2021-08-02T09:00:00Z", "datacontenttype" : "application/cloudevents+json", "data" : {"orderId": "100"}}'
```

Response:
```
INFO[0083] HTTP API Called: POST /v1.0/publish/orderpubsub/orders  app_id=order-processor instance=Yashs-MacBook-Pro.local scope=dapr.runtime.http-info type=log ver=edge
== APP == 127.0.0.1 - - [23/Sep/2022 11:59:30] "POST /orders HTTP/1.1" 200 -
== APP == Subscriber received : {'attributes': {'specversion': '1.0', 'id': 'someCloudEventId', 'source': 'testcloudeventspubsub', 'type': 'com.dapr.cloudevent.sent', 'datacontenttype': 'application/cloudevents+json', 'subject': 'Cloud Events Test', 'time': '2021-08-02T09:00:00Z', 'pubsubname': 'orderpubsub', 'topic': 'orders', 'traceid': '00-b2f1362d4c60fbdfac698deafe3ac630-6161e8ae9292906b-01', 'traceparent': '00-b2f1362d4c60fbdfac698deafe3ac630-6161e8ae9292906b-01', 'tracestate': ''}, 'data': {'orderId': '100'}}
```

Request:
```
curl -X POST [http://localhost:58257/v1.0/publish/orderpubsub/orders](http://localhost:58257/v1.0/publish/orderpubsub/orders) -H "Content-Type: application/cloudevents+json" -d '{"specversion" : "1.0", "type" : "com.dapr.cloudevent.sent", "source" : "testcloudeventspubsub", "subject" : "Cloud Events Test", "id" : "someCloudEventId", "time" : "2021-08-02T09:00:00Z", "datacontenttype" : "application/cloudevents+json", "data" : {"orderId": "100"}}'
```

Response:
```
INFO[0230] HTTP API Called: POST /v1.0/publish/orderpubsub/orders  app_id=order-processor instance=Yashs-MacBook-Pro.local scope=dapr.runtime.http-info type=log ver=edge
== APP == Subscriber received : {'attributes': {'specversion': '1.0', 'id': 'someCloudEventId', 'source': 'testcloudeventspubsub', 'type': 'com.dapr.cloudevent.sent', 'datacontenttype': 'application/cloudevents+json', 'subject': 'Cloud Events Test', 'time': '2021-08-02T09:00:00Z', 'pubsubname': 'orderpubsub', 'topic': 'orders', 'traceid': '00-aee739e0a87bb4c15545ef2a04cdff2d-71d1686f64bba3df-01', 'traceparent': '00-aee739e0a87bb4c15545ef2a04cdff2d-71d1686f64bba3df-01', 'tracestate': ''}, 'data': {'orderId': '100'}}
== APP == 127.0.0.1 - - [23/Sep/2022 12:01:57] "POST /orders HTTP/1.1" 200 -
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -395,7 +395,8 @@ replace github.com/eclipse/paho.mqtt.golang => github.com/shivamkm07/paho.mqtt.g
 // Uncomment for local development for testing with changes in the components-contrib repository.
 // Don't commit with this uncommented!
 //
-// replace github.com/dapr/components-contrib => ../components-contrib
+replace github.com/dapr/components-contrib => github.com/yash-nisar/components-contrib v0.0.0-20220924004408-3316d75c8261
```
Comment (artursouza) [inline go.mod:398]: Please, remove this and pin latest contrib instead.

File: pkg/grpc/api.go [hunk 2]
```diff
@@ -424,13 +426,38 @@ func (a *api) PublishEvent(ctx context.Context, in *runtimev1pb.PublishEventRequ
 
 	data := body
 
+	var cloudEventInterface interface{}
+	var publishTime string
+
+	if in.DataContentType == customCloudEvent {
+		unmarshalError := json.Unmarshal(body, &cloudEventInterface)
```
Comment (artursouza) [inline pkg/grpc/api.go:433]: Why the extra unmarshal? See how it is done for other attributes - I don't think we need this extra step.

/cc @yaron2

File: pkg/http/api.go [hunk 3]
```diff
@@ -1953,6 +1955,32 @@ func (a *api) onPublish(reqCtx *fasthttp.RequestCtx) {
 
 	data := body
 
+	var cloudEventInterface interface{}
+	var publishTime string
+
+	if contentType == customCloudEvent {
```
Comment (artursouza) [inline pkg/http/api.go:1961]: same.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: pkg/grpc/api.go (hunk 2 above)
Comment (yash-nisar) [reply pkg/grpc/api.go:433]: Fixed !

File: pkg/http/api.go (hunk 3 above)
Comment (yash-nisar) [reply pkg/http/api.go:1961]: Fixed !

File: go.mod (hunk 1 above)
Comment (yash-nisar) [reply go.mod:398]: Yup, working on it

**Pull request closed without merging**